File: java/Apparat/src/com/joa_ebert/apparat/abc/io/AbcInputStream.java
/*
 * This file is part of Apparat.
 * 
 * Apparat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Apparat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with Apparat. If not, see <http://www.gnu.org/licenses/>.
 * 
 * Copyright (C) 2009 Joa Ebert
 * http://www.joa-ebert.com/
 * 
 */

package com.joa_ebert.apparat.abc.io;

import java.io.IOException;
import java.io.InputStream;

/**
 * 
 * @author Joa Ebert
 * 
 */
public final class AbcInputStream extends InputStream
{
	private final InputStream input;

	private long position = 0;
	private long markPosition = 0;

	public AbcInputStream( final InputStream input )
	{
		this.input = input;
	}

	@Override
	public int available() throws IOException
	{
		return input.available();
	}

	@Override
	public void close() throws IOException
	{
		input.close();
	}

	public long getPosition()
	{
		return position;
	}

	@Override
	public synchronized void mark( final int readlimit )
	{
		markPosition = position;

		input.mark( readlimit );
	}

	@Override
	public boolean markSupported()
	{
		return input.markSupported();
	}

	@Override
	public int read() throws IOException
	{
		++position;

		return input.read();
	}

	@Override
	public int read( final byte[] b ) throws IOException
	{
		final int bytesRead = input.read( b );

		position += bytesRead;

		return bytesRead;
	}

	@Override
	public int read( final byte[] b, final int offset, final int length )
			throws IOException
	{
		final int bytesRead = input.read( b, offset, length );

		position += bytesRead;

		return bytesRead;
	}

	public double readD64() throws IOException
	{
		final long first = readU08() | ( readU08() << 8 ) | ( readU08() << 16 )
				| ( readU08() << 24 );
		final long second = readU08() | ( readU08() << 8 ) | ( readU08() << 16 )
				| ( readU08() << 24 );

		return Double.longBitsToDouble( first & 0xFFFFFFFFL | second << 32 );
	}

	public int readS24() throws IOException
	{
		final int byte0 = read();
		final int byte1 = read();
		final int byte2 = read();

		int result = ( byte2 << 0x10 ) | ( byte1 << 0x08 ) | byte0;

		if( 0 != ( result & 0x800000 ) )
		{
			result &= 0x7fffff;
			result -= 0x800000;
		}

		return result;
	}

	public int readS32() throws IOException
	{
		long result = readVariableLength();

		if( 0 != ( result & 0x80000000 ) )
		{
			result &= 0x7fffffff;
			result -= 0x80000000;
		}

		return (int)result;
	}

	public String readString() throws IOException
	{
		final int length = readU30();
		final byte[] bytes = new byte[ length ];

		int offset = 0;

		while( offset < length )
		{
			offset += read( bytes, offset, length - offset );
		}

		return new String( bytes, "UTF8" );
	}

	public int readU08() throws IOException
	{
		return read();
	}

	public int readU16() throws IOException
	{
		final int byte0 = read();
		final int byte1 = read();

		return ( byte1 << 0x08 ) | byte0;
	}

	public int readU30() throws IOException
	{
		return (int)( readVariableLength() & 0x3fffffffL );
	}

	public long readU32() throws IOException
	{
		return readVariableLength();
	}

	private long readVariableLength() throws IOException
	{
		long b = read();
		b &= 0xFF;

		long u32 = b;

		if( !( ( u32 & 0x00000080 ) == 0x00000080 ) )
		{
			return u32;
		}

		b = read();
		b &= 0xFF;

		u32 = u32 & 0x0000007f | b << 7;

		if( !( ( u32 & 0x00004000 ) == 0x00004000 ) )
		{
			return u32;
		}

		b = read();
		b &= 0xFF;

		u32 = u32 & 0x00003fff | b << 14;

		if( !( ( u32 & 0x00200000 ) == 0x00200000 ) )
		{
			return u32;
		}

		b = read();
		b &= 0xFF;

		u32 = u32 & 0x001fffff | b << 21;

		if( !( ( u32 & 0x10000000 ) == 0x10000000 ) )
		{
			return u32;
		}

		b = read();
		b &= 0xFF;

		u32 = u32 & 0x0fffffff | b << 28;

		return u32;
	}

	@Override
	public synchronized void reset() throws IOException
	{
		position = markPosition;

		input.reset();
	}

	@Override
	public long skip( final long n ) throws IOException
	{
		position += n;

		return input.skip( n );
	}
}

File: scala/apparat/src/apparat/abc/AbcInputStream.scala
/*
 * This file is part of Apparat.
 * 
 * Apparat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Apparat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with Apparat. If not, see <http://www.gnu.org/licenses/>.
 * 
 * Copyright (C) 2009 Joa Ebert
 * http://www.joa-ebert.com/
 * 
 */
package apparat.abc

import apparat.utils.IO
import java.io.InputStream

class AbcInputStream(input: InputStream) extends InputStream {
  var pos: Int = 0
  def position = pos
  
  private def decodeInt(): Long = {
    var b: Long = read() & 0xff
    var u32 = b
	
    if(!((u32 & 0x00000080) == 0x00000080)) return u32

    b = read() & 0xff
    u32 = u32 & 0x0000007f | b << 7

    if(!((u32 & 0x00004000) == 0x00004000)) return u32

    b = read() & 0xff
    u32 = u32 & 0x00003fff | b << 14

    if(!((u32 & 0x00200000) == 0x00200000)) return u32

    b = read() & 0xff;
    u32 = u32 & 0x001fffff | b << 21;

    if(!((u32 & 0x10000000) == 0x10000000)) return u32

    b = read() & 0xff
    u32 & 0x0fffffff | b << 28;
  }
  
  def readD64() = {
    val first: Long = read() | (read() << 8) | (read() << 16) | (read() << 24);
    val second: Long = read() | (read() << 8) | (read() << 16) | (read() << 24);
    java.lang.Double.longBitsToDouble(first & 0xFFFFFFFFL | second << 32);
  }
  
  def readS24() = {
    val b0 = read
    val b1 = read
    val r = (read() << 0x10) | (b1 << 0x08) | b0;

    if(0 != (r & 0x800000)) ((r & 0x7fffff) - 0x800000)
    else r
  }
  
  def readS32(): Int = {
    val r = decodeInt
    if(0 != (r & 0x80000000)) ((r & 0x7fffffff) - 0x80000000).asInstanceOf[Int]
    else r.asInstanceOf[Int]
  }
  
  def readString() = new String(IO.read(readU30)(this), "UTF8")
  
  def readU08() = read

  def readU16() = {
    val b0 = read
    ( read() << 0x08 ) | b0
  }
  
  def readU30() = (decodeInt & 0x3fffffffL).asInstanceOf[Int]
  
  def readU32() = decodeInt
  
  override def available() = input available
  override def close() = input close()
  override def read() = {
    pos += 1
    input read()
  }
  override def read(b: Array[Byte]) = {
    val n = input read(b)
    pos += n
    n
  }
  override def read(b: Array[Byte], off: Int, len: Int) = {
    val n = input read(b, off, len)
    pos += n
    n
  }
  override def reset() = input reset()
  override def skip(n: Long) = {
    pos += n.asInstanceOf[Int]
    input skip n
  }
}
